{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# transform `pseudogenes.csv` in gff3 \n",
    "\n",
    "This is needed to standardise the addition of attributes at the end of the screening for core and dispensable genes, so that one of the input will be always an annotation file in gff3 format (ncbi annotations are in gff3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('/home/lia/A_pangenome')\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome  start   end     coverage  Genewise_score  position    size  Strand  Type  Pater        Species     Stop_codon  Frameshift\n",
      "Chr1        6144    6269    0.09      56.39           UTR         129   plus    FRAG  AT1G01010.1  A.thaliana  0           0\n",
      "Chr1        47087   47325   0.06      74.09           Intergenic  206   plus    FRAG  AT1G02730.1  A.thaliana  1           3\n",
      "Chr1        89932   90153   0.02      24.75           Intergenic  51    plus    FRAG  AT5G41740.2  A.thaliana  1           3\n",
      "Chr1        259011  259151  0.06      48.95           Intergenic  84    plus    FRAG  AT1G01690.1  A.thaliana  0           0\n",
      "Chr1        266562  267495  0.99      68.96           UTR         979   plus    DUP   AT4G00525.1  A.thaliana  0           0\n",
      "Chr1        426018  426176  0.06      17.28           Intergenic  36    plus    FRAG  AT3G04430.1  A.thaliana  1           1\n",
      "Chr1        578305  578484  0.04      47.13           Intron_cds  102   plus    FRAG  AT1G05120.1  A.thaliana  0           0\n",
      "Chr1        582401  582585  0.46      52.69           Intergenic  236   plus    SE    AT4G02160.1  A.thaliana  0           1\n",
      "Chr1        582634  582971  0.57      117.33          Intergenic  247   plus    SE    AT5G61710.1  A.thaliana  0           1\n"
     ]
    }
   ],
   "source": [
    "!head pseudogenes.csv | column -t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>coverage</th>\n",
       "      <th>Genewise_score</th>\n",
       "      <th>position</th>\n",
       "      <th>size</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Type</th>\n",
       "      <th>Pater</th>\n",
       "      <th>Species</th>\n",
       "      <th>Stop_codon</th>\n",
       "      <th>Frameshift</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>6144</td>\n",
       "      <td>6269</td>\n",
       "      <td>0.09</td>\n",
       "      <td>56.39</td>\n",
       "      <td>UTR</td>\n",
       "      <td>129</td>\n",
       "      <td>plus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT1G01010.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>47087</td>\n",
       "      <td>47325</td>\n",
       "      <td>0.06</td>\n",
       "      <td>74.09</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>206</td>\n",
       "      <td>plus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT1G02730.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>89932</td>\n",
       "      <td>90153</td>\n",
       "      <td>0.02</td>\n",
       "      <td>24.75</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>51</td>\n",
       "      <td>plus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT5G41740.2</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>259011</td>\n",
       "      <td>259151</td>\n",
       "      <td>0.06</td>\n",
       "      <td>48.95</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>84</td>\n",
       "      <td>plus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT1G01690.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>266562</td>\n",
       "      <td>267495</td>\n",
       "      <td>0.99</td>\n",
       "      <td>68.96</td>\n",
       "      <td>UTR</td>\n",
       "      <td>979</td>\n",
       "      <td>plus</td>\n",
       "      <td>DUP</td>\n",
       "      <td>AT4G00525.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   start     end  coverage Genewise_score    position  size  \\\n",
       "0       Chr1    6144    6269      0.09          56.39         UTR   129   \n",
       "1       Chr1   47087   47325      0.06          74.09  Intergenic   206   \n",
       "2       Chr1   89932   90153      0.02          24.75  Intergenic    51   \n",
       "3       Chr1  259011  259151      0.06          48.95  Intergenic    84   \n",
       "4       Chr1  266562  267495      0.99          68.96         UTR   979   \n",
       "\n",
       "  Strand  Type        Pater     Species  Stop_codon  Frameshift  \n",
       "0   plus  FRAG  AT1G01010.1  A.thaliana           0           0  \n",
       "1   plus  FRAG  AT1G02730.1  A.thaliana           1           3  \n",
       "2   plus  FRAG  AT5G41740.2  A.thaliana           1           3  \n",
       "3   plus  FRAG  AT1G01690.1  A.thaliana           0           0  \n",
       "4   plus   DUP  AT4G00525.1  A.thaliana           0           0  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('pseudogenes.csv', sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>coverage</th>\n",
       "      <th>Genewise_score</th>\n",
       "      <th>position</th>\n",
       "      <th>size</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Type</th>\n",
       "      <th>Pater</th>\n",
       "      <th>Species</th>\n",
       "      <th>Stop_codon</th>\n",
       "      <th>Frameshift</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>score</th>\n",
       "      <th>score2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>6144</td>\n",
       "      <td>6269</td>\n",
       "      <td>0.09</td>\n",
       "      <td>56.39</td>\n",
       "      <td>UTR</td>\n",
       "      <td>129</td>\n",
       "      <td>plus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT1G01010.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>47087</td>\n",
       "      <td>47325</td>\n",
       "      <td>0.06</td>\n",
       "      <td>74.09</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>206</td>\n",
       "      <td>plus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT1G02730.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>89932</td>\n",
       "      <td>90153</td>\n",
       "      <td>0.02</td>\n",
       "      <td>24.75</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>51</td>\n",
       "      <td>plus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT5G41740.2</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>259011</td>\n",
       "      <td>259151</td>\n",
       "      <td>0.06</td>\n",
       "      <td>48.95</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>84</td>\n",
       "      <td>plus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT1G01690.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>266562</td>\n",
       "      <td>267495</td>\n",
       "      <td>0.99</td>\n",
       "      <td>68.96</td>\n",
       "      <td>UTR</td>\n",
       "      <td>979</td>\n",
       "      <td>plus</td>\n",
       "      <td>DUP</td>\n",
       "      <td>AT4G00525.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4670</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>610030</td>\n",
       "      <td>610113</td>\n",
       "      <td>0.04</td>\n",
       "      <td>15.65</td>\n",
       "      <td>Intron_cds</td>\n",
       "      <td>51</td>\n",
       "      <td>minus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT3G28410.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4671</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>489952</td>\n",
       "      <td>491331</td>\n",
       "      <td>0.11</td>\n",
       "      <td>91.26</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>1350</td>\n",
       "      <td>minus</td>\n",
       "      <td>DUP</td>\n",
       "      <td>AT5G02320.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4672</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>240395</td>\n",
       "      <td>240439</td>\n",
       "      <td>0.03</td>\n",
       "      <td>52.75</td>\n",
       "      <td>Intron_cds</td>\n",
       "      <td>105</td>\n",
       "      <td>minus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT4G00020.2</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4673</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>238426</td>\n",
       "      <td>238539</td>\n",
       "      <td>0.02</td>\n",
       "      <td>38.62</td>\n",
       "      <td>Intron_cds</td>\n",
       "      <td>69</td>\n",
       "      <td>minus</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT4G00020.2</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4674</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>183187</td>\n",
       "      <td>183285</td>\n",
       "      <td>0.40</td>\n",
       "      <td>none</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>135</td>\n",
       "      <td>minus</td>\n",
       "      <td>SE</td>\n",
       "      <td>AT5G01445.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4675 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome   start     end  coverage Genewise_score    position  size  \\\n",
       "0          Chr1    6144    6269      0.09          56.39         UTR   129   \n",
       "1          Chr1   47087   47325      0.06          74.09  Intergenic   206   \n",
       "2          Chr1   89932   90153      0.02          24.75  Intergenic    51   \n",
       "3          Chr1  259011  259151      0.06          48.95  Intergenic    84   \n",
       "4          Chr1  266562  267495      0.99          68.96         UTR   979   \n",
       "...         ...     ...     ...       ...            ...         ...   ...   \n",
       "4670       Chr5  610030  610113      0.04          15.65  Intron_cds    51   \n",
       "4671       Chr5  489952  491331      0.11          91.26  Intergenic  1350   \n",
       "4672       Chr5  240395  240439      0.03          52.75  Intron_cds   105   \n",
       "4673       Chr5  238426  238539      0.02          38.62  Intron_cds    69   \n",
       "4674       Chr5  183187  183285      0.40           none  Intergenic   135   \n",
       "\n",
       "     Strand  Type        Pater     Species  Stop_codon  Frameshift  \\\n",
       "0      plus  FRAG  AT1G01010.1  A.thaliana           0           0   \n",
       "1      plus  FRAG  AT1G02730.1  A.thaliana           1           3   \n",
       "2      plus  FRAG  AT5G41740.2  A.thaliana           1           3   \n",
       "3      plus  FRAG  AT1G01690.1  A.thaliana           0           0   \n",
       "4      plus   DUP  AT4G00525.1  A.thaliana           0           0   \n",
       "...     ...   ...          ...         ...         ...         ...   \n",
       "4670  minus  FRAG  AT3G28410.1  A.thaliana           0           0   \n",
       "4671  minus   DUP  AT5G02320.1  A.thaliana           0           0   \n",
       "4672  minus  FRAG  AT4G00020.2  A.thaliana           0           0   \n",
       "4673  minus  FRAG  AT4G00020.2  A.thaliana           0           0   \n",
       "4674  minus    SE  AT5G01445.1  A.thaliana           1           1   \n",
       "\n",
       "           source        type score score2  \n",
       "0     bed_to_gff3  pseudogene     .      .  \n",
       "1     bed_to_gff3  pseudogene     .      .  \n",
       "2     bed_to_gff3  pseudogene     .      .  \n",
       "3     bed_to_gff3  pseudogene     .      .  \n",
       "4     bed_to_gff3  pseudogene     .      .  \n",
       "...           ...         ...   ...    ...  \n",
       "4670  bed_to_gff3  pseudogene     .      .  \n",
       "4671  bed_to_gff3  pseudogene     .      .  \n",
       "4672  bed_to_gff3  pseudogene     .      .  \n",
       "4673  bed_to_gff3  pseudogene     .      .  \n",
       "4674  bed_to_gff3  pseudogene     .      .  \n",
       "\n",
       "[4675 rows x 17 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"source\"] = \"bed_to_gff3\"\n",
    "df[\"type\"] = \"pseudogene\"\n",
    "df[\"score\"] = \".\"\n",
    "df[\"score2\"] = \".\"\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>score2</th>\n",
       "      <th>coverage</th>\n",
       "      <th>Genewise_score</th>\n",
       "      <th>position</th>\n",
       "      <th>size</th>\n",
       "      <th>Type</th>\n",
       "      <th>Pater</th>\n",
       "      <th>Species</th>\n",
       "      <th>Stop_codon</th>\n",
       "      <th>Frameshift</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>6144</td>\n",
       "      <td>6269</td>\n",
       "      <td>.</td>\n",
       "      <td>plus</td>\n",
       "      <td>.</td>\n",
       "      <td>0.09</td>\n",
       "      <td>56.39</td>\n",
       "      <td>UTR</td>\n",
       "      <td>129</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT1G01010.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>47087</td>\n",
       "      <td>47325</td>\n",
       "      <td>.</td>\n",
       "      <td>plus</td>\n",
       "      <td>.</td>\n",
       "      <td>0.06</td>\n",
       "      <td>74.09</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>206</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT1G02730.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>89932</td>\n",
       "      <td>90153</td>\n",
       "      <td>.</td>\n",
       "      <td>plus</td>\n",
       "      <td>.</td>\n",
       "      <td>0.02</td>\n",
       "      <td>24.75</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>51</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT5G41740.2</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>259011</td>\n",
       "      <td>259151</td>\n",
       "      <td>.</td>\n",
       "      <td>plus</td>\n",
       "      <td>.</td>\n",
       "      <td>0.06</td>\n",
       "      <td>48.95</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>84</td>\n",
       "      <td>FRAG</td>\n",
       "      <td>AT1G01690.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>266562</td>\n",
       "      <td>267495</td>\n",
       "      <td>.</td>\n",
       "      <td>plus</td>\n",
       "      <td>.</td>\n",
       "      <td>0.99</td>\n",
       "      <td>68.96</td>\n",
       "      <td>UTR</td>\n",
       "      <td>979</td>\n",
       "      <td>DUP</td>\n",
       "      <td>AT4G00525.1</td>\n",
       "      <td>A.thaliana</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome       source        type   start     end score Strand score2  \\\n",
       "0       Chr1  bed_to_gff3  pseudogene    6144    6269     .   plus      .   \n",
       "1       Chr1  bed_to_gff3  pseudogene   47087   47325     .   plus      .   \n",
       "2       Chr1  bed_to_gff3  pseudogene   89932   90153     .   plus      .   \n",
       "3       Chr1  bed_to_gff3  pseudogene  259011  259151     .   plus      .   \n",
       "4       Chr1  bed_to_gff3  pseudogene  266562  267495     .   plus      .   \n",
       "\n",
       "   coverage Genewise_score    position  size  Type        Pater     Species  \\\n",
       "0      0.09          56.39         UTR   129  FRAG  AT1G01010.1  A.thaliana   \n",
       "1      0.06          74.09  Intergenic   206  FRAG  AT1G02730.1  A.thaliana   \n",
       "2      0.02          24.75  Intergenic    51  FRAG  AT5G41740.2  A.thaliana   \n",
       "3      0.06          48.95  Intergenic    84  FRAG  AT1G01690.1  A.thaliana   \n",
       "4      0.99          68.96         UTR   979   DUP  AT4G00525.1  A.thaliana   \n",
       "\n",
       "   Stop_codon  Frameshift  \n",
       "0           0           0  \n",
       "1           1           3  \n",
       "2           1           3  \n",
       "3           0           0  \n",
       "4           0           0  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rearrange the columns (not strictly necessary, but I like to have the assembly_name column first)\n",
    "# Get a list of all the column names\n",
    "cols = list(df.columns)\n",
    "# Remove 'assembly_name' from the list\n",
    "cols.remove('Chromosome')\n",
    "cols.remove('source')\n",
    "cols.remove('type')\n",
    "cols.remove('start')\n",
    "cols.remove('end')\n",
    "cols.remove('score')\n",
    "cols.remove('Strand')\n",
    "cols.remove('score2')\n",
    "# Create a new list with 'assembly_name' first, followed by the other columns\n",
    "cols = ['Chromosome', 'source', 'type', 'start', 'end', 'score', 'Strand', 'score2'] + cols\n",
    "# Reorder the DataFrame\n",
    "df = df[cols]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_5118/3884775361.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['attributes'] = df[cols_to_join].apply(lambda row: ';'.join(f'{col}={row[col]}' for col in cols_to_join), axis=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>score2</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP002684.1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>6144</td>\n",
       "      <td>6269</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>coverage=0.09;Genewise_score=56.39;position=UT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP002684.1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>47087</td>\n",
       "      <td>47325</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>coverage=0.06;Genewise_score=74.09;position=In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP002684.1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>89932</td>\n",
       "      <td>90153</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>coverage=0.02;Genewise_score=24.75;position=In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP002684.1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>259011</td>\n",
       "      <td>259151</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>coverage=0.06;Genewise_score=48.95;position=In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP002684.1</td>\n",
       "      <td>bed_to_gff3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>266562</td>\n",
       "      <td>267495</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>coverage=0.99;Genewise_score=68.96;position=UT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome       source        type   start     end score Strand score2  \\\n",
       "0  CP002684.1  bed_to_gff3  pseudogene    6144    6269     .      +      .   \n",
       "1  CP002684.1  bed_to_gff3  pseudogene   47087   47325     .      +      .   \n",
       "2  CP002684.1  bed_to_gff3  pseudogene   89932   90153     .      +      .   \n",
       "3  CP002684.1  bed_to_gff3  pseudogene  259011  259151     .      +      .   \n",
       "4  CP002684.1  bed_to_gff3  pseudogene  266562  267495     .      +      .   \n",
       "\n",
       "                                          attributes  \n",
       "0  coverage=0.09;Genewise_score=56.39;position=UT...  \n",
       "1  coverage=0.06;Genewise_score=74.09;position=In...  \n",
       "2  coverage=0.02;Genewise_score=24.75;position=In...  \n",
       "3  coverage=0.06;Genewise_score=48.95;position=In...  \n",
       "4  coverage=0.99;Genewise_score=68.96;position=UT...  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Specify the columns to be joined\n",
    "cols_to_join = ['coverage', 'Genewise_score', 'position', 'size', 'Type', 'Pater', 'Species', 'Stop_codon', 'Frameshift']\n",
    "\n",
    "# Create the new column\n",
    "df['attributes'] = df[cols_to_join].apply(lambda row: ';'.join(f'{col}={row[col]}' for col in cols_to_join), axis=1)\n",
    "\n",
    "# Drop the original columns\n",
    "df = df.drop(columns=cols_to_join)\n",
    "\n",
    "# substitute plus and minus with + and -\n",
    "df['Strand'] = df['Strand'].replace({'plus': '+', 'minus': '-'})\n",
    "\n",
    "#change the name of the chromosome to adapt them to those used in the pangenome\n",
    "df['Chromosome'] = df['Chromosome'].replace({'Chr1': 'CP002684.1', 'Chr2': 'CP002685.1' , 'Chr3': 'CP002686.1', 'Chr4': 'CP002687.1', 'Chr5': 'CP002688.1'})\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CP002684.1' 'CP002685.1' 'CP002686.1' 'CP002687.1' 'CP002688.1']\n"
     ]
    }
   ],
   "source": [
    "print(df['Chromosome'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##gff-version             3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \n",
      "#!gff-spec-version        1.21                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n",
      "#!processor               NCBI                                                              annotwriter                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
      "#!genome-build            TAIR10.1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
      "#!genome-build-accession  NCBI_Assembly:GCA_000001735.2                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
      "##sequence-region         CP002684.1                                                        1            30427671                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
      "##species                 https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=3702                                                                                                                                                                                                                                                                                                                                                                                                                                         \n",
      "CP002684.1                Genbank                                                           region       1         30427671  .  +  .  ID=CP002684.1:1..30427671;Dbxref=taxon:3702;Name=1;chromosome=1;ecotype=Columbia;gbkey=Src;mol_type=genomic                                        DNA                                                                                                                                                                                                                                       \n",
      "CP002684.1                Genbank                                                           gene         3631      5899      .  +  .  ID=gene-AT1G01010;Dbxref=Araport:AT1G01010,TAIR:AT1G01010;Name=NAC001;gbkey=Gene;gene=NAC001;gene_biotype=protein_coding;gene_synonym=ANAC001,NAC  domain  containing  protein      1,T25K16.1,T25K16_1;locus_tag=AT1G01010                                                                                                                                                                  \n",
      "CP002684.1                Genbank                                                           mRNA         3631      5899      .  +  .  ID=rna-gnl|JCVI|mRNA.AT1G01010.1;Parent=gene-AT1G01010;Dbxref=Araport:AT1G01010,TAIR:AT1G01010;gbkey=mRNA;gene=NAC001;inference=similar            to      RNA         sequence%2C  mRNA:INSD:BT001115.1%2CINSD:AF439834.1%2CINSD:AK226863.1;locus_tag=AT1G01010;orig_protein_id=gnl|JCVI|AT1G01010.1;orig_transcript_id=gnl|JCVI|mRNA.AT1G01010.1;product=NAC  domain  containing  protein  1\n"
     ]
    }
   ],
   "source": [
    "!head GCA_000001735.2_TAIR10.1_genomic.gff | column -t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('pseudogenes.gff', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CP002684.1  bed_to_gff3  pseudogene  6144    6269    .  +  .  coverage=0.09;Genewise_score=56.39;position=UTR;size=129;Type=FRAG;Pater=AT1G01010.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002684.1  bed_to_gff3  pseudogene  47087   47325   .  +  .  coverage=0.06;Genewise_score=74.09;position=Intergenic;size=206;Type=FRAG;Pater=AT1G02730.1;Species=A.thaliana;Stop_codon=1;Frameshift=3\n",
      "CP002684.1  bed_to_gff3  pseudogene  89932   90153   .  +  .  coverage=0.02;Genewise_score=24.75;position=Intergenic;size=51;Type=FRAG;Pater=AT5G41740.2;Species=A.thaliana;Stop_codon=1;Frameshift=3\n",
      "CP002684.1  bed_to_gff3  pseudogene  259011  259151  .  +  .  coverage=0.06;Genewise_score=48.95;position=Intergenic;size=84;Type=FRAG;Pater=AT1G01690.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002684.1  bed_to_gff3  pseudogene  266562  267495  .  +  .  coverage=0.99;Genewise_score=68.96;position=UTR;size=979;Type=DUP;Pater=AT4G00525.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002684.1  bed_to_gff3  pseudogene  426018  426176  .  +  .  coverage=0.06;Genewise_score=17.28;position=Intergenic;size=36;Type=FRAG;Pater=AT3G04430.1;Species=A.thaliana;Stop_codon=1;Frameshift=1\n",
      "CP002684.1  bed_to_gff3  pseudogene  578305  578484  .  +  .  coverage=0.04;Genewise_score=47.13;position=Intron_cds;size=102;Type=FRAG;Pater=AT1G05120.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002684.1  bed_to_gff3  pseudogene  582401  582585  .  +  .  coverage=0.46;Genewise_score=52.69;position=Intergenic;size=236;Type=SE;Pater=AT4G02160.1;Species=A.thaliana;Stop_codon=0;Frameshift=1\n",
      "CP002684.1  bed_to_gff3  pseudogene  582634  582971  .  +  .  coverage=0.57;Genewise_score=117.33;position=Intergenic;size=247;Type=SE;Pater=AT5G61710.1;Species=A.thaliana;Stop_codon=0;Frameshift=1\n",
      "CP002684.1  bed_to_gff3  pseudogene  893047  893493  .  +  .  coverage=0.13;Genewise_score=45.73;position=Intergenic;size=177;Type=RET;Pater=AT2G42460.1;Species=A.thaliana;Stop_codon=1;Frameshift=1\n"
     ]
    }
   ],
   "source": [
    "!head pseudogenes.gff | column -t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CP002688.1  bed_to_gff3  pseudogene  1341997  1342098  .  -  .  coverage=0.04;Genewise_score=19.96;position=Intergenic;size=51;Type=FRAG;Pater=AT3G23280.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002688.1  bed_to_gff3  pseudogene  1031516  1032032  .  -  .  coverage=0.4;Genewise_score=221.81;position=Intergenic;size=460;Type=FRAG;Pater=AT5G03960.1;Species=A.thaliana;Stop_codon=3;Frameshift=4\n",
      "CP002688.1  bed_to_gff3  pseudogene  911089   911463   .  -  .  coverage=0.99;Genewise_score=90.11;position=Intergenic;size=379;Type=SE;Pater=AT5G38440.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002688.1  bed_to_gff3  pseudogene  706016   706120   .  -  .  coverage=0.29;Genewise_score=67.74;position=Intergenic;size=99;Type=FRAG;Pater=AT5G03030.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002688.1  bed_to_gff3  pseudogene  614386   614598   .  -  .  coverage=0.31;Genewise_score=89.77;position=UTR;size=207;Type=FRAG;Pater=AT3G53490.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002688.1  bed_to_gff3  pseudogene  610030   610113   .  -  .  coverage=0.04;Genewise_score=15.65;position=Intron_cds;size=51;Type=FRAG;Pater=AT3G28410.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002688.1  bed_to_gff3  pseudogene  489952   491331   .  -  .  coverage=0.11;Genewise_score=91.26;position=Intergenic;size=1350;Type=DUP;Pater=AT5G02320.1;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002688.1  bed_to_gff3  pseudogene  240395   240439   .  -  .  coverage=0.03;Genewise_score=52.75;position=Intron_cds;size=105;Type=FRAG;Pater=AT4G00020.2;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002688.1  bed_to_gff3  pseudogene  238426   238539   .  -  .  coverage=0.02;Genewise_score=38.62;position=Intron_cds;size=69;Type=FRAG;Pater=AT4G00020.2;Species=A.thaliana;Stop_codon=0;Frameshift=0\n",
      "CP002688.1  bed_to_gff3  pseudogene  183187   183285   .  -  .  coverage=0.4;Genewise_score=none;position=Intergenic;size=135;Type=SE;Pater=AT5G01445.1;Species=A.thaliana;Stop_codon=1;Frameshift=1\n"
     ]
    }
   ],
   "source": [
    "!tail pseudogenes.gff | column -t"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
